### PR TITLE
Support EP with HSDP

### PR DESCRIPTION
**Pull request by wplf**
# What does this PR do ?

dev PR: https://github.com/NVIDIA/Megatron-LM/pull/2800

This MR adds HSDP support for Expert Parallelism. With this change, DeepSeek‑v3 and other MoE models can be trained using M‑FSDP as well as HSDP+EP. HSDP enables better use of local bandwidth within a larger nvlink domain.

We ran convergence tests on the [DeepSeek Proxy model](https://gitlab-master.nvidia.com/arch_moe_exploration/megatron-moe-scripts/-/blob/master/model_configs/benchmarking/DeepSeek-V3-Proxy.yaml?ref_type=heads). model. The experiments were based on the latest dev branch, with an additional [patch](https://github.com/xuwchen/Megatron-LM/commit/2d8a285a4964202567f178f1c0f2e6ab25aa1a54) that fixes MoE checkpoint save/load. 

To validate the correctness of HSDP+EP, we loaded an FSDP+EP checkpoint at step 100 and resumed training with HSDP+EP. The loss curves and gradient norms matched closely.

image
image

f748c47e86d949904ac404d41dd1506c

For example, to use HSDP=2 and EP size=2, you need to add these argument to your script 
```bash
--num-distributed-optimizer-instances 2 \
--outer-dp-sharding-strategy optim
--expert-model-parallel-size 2 
```

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @megatron-oncall.

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

Feel free to message or comment the @megatron-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

**Comment by shjwudp** [comment]
Hi @wplf , could you please fix the lint issue and synchronize the code with the [PR-2800](https://github.com/NVIDIA/Megatron-LM/pull/2800)? thanks.

**Comment by wplf** [comment]
> Hi @wplf , could you please fix the lint issue and synchronize the code with the [PR-2800](https://github.com/NVIDIA/Megatron-LM/pull/2800)? thanks.

OKK, I'll do it right now.

**Comment by wplf** [comment]
Hi jack，@shjwudp.
I've synchronized this pr with [pr-2800](https://github.com/NVIDIA/Megatron-LM/pull/2800).
Please tell me if anything else I can do.

File: megatron/core/distributed/fsdp/src/README.md [hunk 1]
```diff
@@ -150,18 +150,19 @@ device_mesh = torch.distributed.device_mesh.init_device_mesh(
 # This sub-mesh can be provided to distributed samplers or dataloaders.
 device_mesh[("dp_outer", "dp_shard")]._flatten("dp")
 # Only required if using CP. Otherwise, just pass dp_shard to FSDP.
-device_mesh[("dp_shard", "cp")]._flatten("dp_shard_cp")
+device_mesh[("dp_shard", "cp")]._flatten("dp_cp")
```
Comment (cspades) [inline megatron/core/distributed/fsdp/src/README.md]: Nit: Can we revert all of these "dp_cp" naming changes, and only in the `README.md`? Would like to be clear in the documentation about DP-Outer and DP-Shard, and this is definitely a inner DP-Shard group.

MCore adapter is fine, since there is a clear naming of intra-DP and inter-DP, but for `fully_shard` would like to match the conventional names matching TorchTitan and PyTorch suggestions. (Basically, just don't want to use only "DP" it should be DP-something, not the full DP group.)

**Comment by cspades** [comment]
Also, any way we can add a functional test into MCore representing EP + HFSDP? 👀

File: megatron/core/distributed/fsdp/mcore_fsdp_adapter.py [hunk 2]
```diff
@@ -243,6 +254,18 @@ def _init_dist_index(self, pg_collection):
...
         if enable_hsdp:
+            if expt_dp_group is not None:
+                expt_mesh = _get_hsdp_tp_mesh(
+                    outer_fsdp_group, expt_dp_group, expt_tp_group, ep_size=ep_group.size()
+                )
+                expt_device_mesh = DeviceMesh.from_group(
+                    [outer_fsdp_group, expt_dp_group, expt_tp_group],
+                    device_type="cuda",
+                    mesh=expt_mesh.tolist(),
+                    mesh_dim_names=["outer_fsdp_dp", "dp_cp", "tp"],
```
Comment (BoxiangW) [inline megatron/core/distributed/fsdp/mcore_fsdp_adapter.py:265]: Agree with Cory, this naming feels weird. Before we used 'dp_replicate' and 'dp_shard' for outer and inner here. This is aligned with FSDP2's naming

Comment (wplf) [reply megatron/core/distributed/fsdp/mcore_fsdp_adapter.py:265]: Hi, Boxiang.
I defined the HSDP+EP mesh name as mesh_dim_names=["outer_fsdp_dp", "dp_cp", "tp"], following the HSDP mesh definition in the https://github.com/NVIDIA/Megatron-LM/blob/30dea5d8f0213d2799911bafef801bbdfadd8f17/megatron/core/distributed/fsdp/mcore_fsdp_adapter.py#L253.
Going with `dp_replicate` would require changing the mesh names for both `hsdp_ep` and `hsdp`. It could be many changes. 
What are your thoughts？ @shjwudp @cspades

File: megatron/core/distributed/fsdp/src/README.md (hunk 1 above)
Comment (wplf) [reply megatron/core/distributed/fsdp/src/README.md]: Hi, cory. I've reverted `dp_cp`.

Please let me know if there's anything I can do.

**Comment by wplf** [comment]
> Also, any way we can add a functional test into MCore representing EP + HFSDP? 👀

OK, I'll try to add a MOE model in `Megatron-LM/tests/unit_tests/distributed/fsdp/test_mfsdp_fully_shard.py`.

File: megatron/core/distributed/fsdp/mcore_fsdp_adapter.py (hunk 2 above)
Comment (wplf) [reply megatron/core/distributed/fsdp/mcore_fsdp_adapter.py:265]: FYI, the `dp_replica degree `will replicate optimizer and weight and gradient.
But in megatron, `outer_fsdp_dp` can 
1. replicate them both or 
2. replicate  weight and gradient only and shard optimizer state, which is not implemented by torchtitan as far as I know.

So it's better not to use `dp_replica` in my opinion.

Comment (cspades) [reply megatron/core/distributed/fsdp/mcore_fsdp_adapter.py:265]: I just synced with @BoxiangW on this today.

First of all, I'm terribly sorry about the variable and dimension naming in Megatron-FSDP in general - it's not always easily understandable, and I ended up setting a lot of imaginary/artificial constraints about what I wanted to call things, e.g. "dp-shard" to match FSDP2, and "dp-outer" because it's not always vanilla HSDP with replicated params (although technically, even HFSDP with optimizer full-sharding does replicate compute params, just not main params or the optimizer state, but I digress).

That being said, let's just go with the current "outer" and "shard" naming (basically, just follow the existing naming conventions reused all over the place in the documentation and code) and in a future PR we can discuss the "best" names for stuff and make sweeping changes if users really hate the variable / argument names.

Let's not fix what isn't broke (yet). 🙏🏻

**Comment by wplf** [comment]
> > Also, any way we can add a functional test into MCore representing EP + HFSDP? 👀
> 
> OK, I'll try to add a MOE model in `Megatron-LM/tests/unit_tests/distributed/fsdp/test_mfsdp_fully_shard.py`.

Hi, @cspades @yanring.

Jack just add an unittest for MFSDP to main branch for HSDP + EP in https://github.com/NVIDIA/Megatron-LM/blob/9f05aac7dcd52ced62d72dd9acc933bf23770660/tests/unit_tests/distributed/megatron_fsdp/test_mcore_fully_sharded_data_parallel.py#L635.
I've merged it, already.

Both `no_shard` and `optim` for HSDP in `outer-dp-sharding-strategy` is tested.
The result of HSDP+EP looks fine.

image


Thanks to Jack's efforts.  cc @shjwudp 
Please let me know if there if anything else I need to do to merge this PR.

**Review by shjwudp — APPROVED** [review]
LGTM, thank you!

**Review by jaredcasper — APPROVED** [review]
